### PR TITLE
TransactionQueue

### PR DESCRIPTION
**Pull request by tomusdrw**
Still work in progress but needs some eyeballs.

### CODE REVIEW

File: util/src/executor.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+				self.$field_.eq(&other.$field_)
+			}
+		}
+	}
+}
+
+/// Some work to be done in a thread pool
+pub trait Task: Ord + Send + 'static {
+	/// Task result
+	type TResult: Send + 'static;
```
Comment (gavofyork) [inline util/src/executor.rs]: avoid this convention for naming. gav's rules of naming number 1: make names pronouncable.

File: util/src/executor.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+		self.finished.store(true, atomic::Ordering::Relaxed);
+		self.wait.notify_all();
+
+		for thread in self.threads.drain(..) {
+			thread.join().unwrap();
+		}
+	}
+}
+
+impl<TTask, TQueue> Executor<TTask> for ThreadPoolExecutor<TTask, TQueue>
```
Comment (gavofyork) [inline util/src/executor.rs]: is this rust convention for naming generic parameters?

**Comment by gavofyork** [comment]
looks good so far.

**Comment by NikVolf** [comment]
is it 'in progress'? or 'need review'? :)

File: ethcore/src/transaction.rs [hunk 3]
```diff
@@ -125,8 +125,8 @@ impl Transaction {
 			r: r,
 			s: s,
 			v: v + 27,
-			hash: RefCell::new(None),
-			sender: RefCell::new(None)
+			hash: RwLock::new(None),
```
Comment (gavofyork) [inline ethcore/src/transaction.rs]: is `Transaction` really so substantial that references to it need to be thread-safe?

Comment (arkpar) [inline ethcore/src/transaction.rs]: +1 Locking should be moved to containing object

Comment (arkpar) [inline ethcore/src/transaction.rs]: Also consider using `Cell` instead of `RefCell` here

File: util/src/executor.rs [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+use std::boxed::FnBox;
+use std::cmp::{Ord, Ordering};
+use std::collections::binary_heap::BinaryHeap;
+use std::marker::PhantomData;
+use std::sync::{atomic, Arc, Condvar, Mutex};
+use std::thread::{JoinHandle, spawn};
+
+#[macro_export]
+macro_rules! trivial_ordering {
```
Comment (arkpar) [inline util/src/executor.rs]: This macro is only used once?

File: sync/src/chain.rs [hunk 5]
```diff
@@ -881,8 +891,17 @@ impl ChainSync {
...
 		}
 	}
+
 	/// Called when peer sends us new transactions
-	fn on_peer_transactions(&mut self, _io: &mut SyncIo, _peer_id: PeerId, _r: &UntrustedRlp) -> Result<(), PacketDecodeError> {
+	fn on_peer_transactions(&mut self, io: &mut SyncIo, peer_id: PeerId, r: &UntrustedRlp) -> Result<(), PacketDecodeError> {
+		let chain = io.chain();
+		let item_count = r.item_count();
+		trace!(target: "sync", "{} -> Transactions ({} entries)", peer_id, item_count);
+		let fetch_latest_nonce = |a : &Address| chain.nonce(a);
```
Comment (tomusdrw) [inline sync/src/chain.rs]: When syncing from start it panics when asking for `nonce` with
```
TrieDB::new(d6d2…d52c): Trie root not found!
```
in: `state::State::from_existing`

I will be investigating what happens there.

File: sync/src/transaction_queue.rs [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+		}
+
+		// and senders
+		let a_sender = self.sender();
+		let b_sender = b.sender();
+		a_sender.cmp(&b_sender)
+	}
+}
+
+struct TxsByPriorityAndAddress {
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: generally dislike using the shorthand "tx" to mean "transaction". i'll let it go here though since it's confined to a single file.

File: sync/src/transaction_queue.rs [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+// TODO [todr] - own transactions should have higher priority
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: also, transactions coming from reverted blocks should have higher priority.

File: sync/src/transaction_queue.rs [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+use util::uint::{Uint, U256};
+use util::hash::{Address};
+use util::table::*;
+use ethcore::transaction::*;
+
+
+#[derive(Clone, Debug)]
+struct VerifiedTransaction {
+	tx: SignedTransaction,
+	nonce_height: U256
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: missing trailing `,`

File: sync/src/transaction_queue.rs [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+}
+
+impl VerifiedTransaction {
+	pub fn new(tx: SignedTransaction, nonce_height: U256) -> VerifiedTransaction {
+		VerifiedTransaction {
+			tx: tx,
+			nonce_height: nonce_height
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: missing trailing `,`. i guess this is endemic in this PR...

File: sync/src/transaction_queue.rs [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+				}
+			}
+			self.future.enforce_limit();
+			return;
+		}
+
+		// Remove from future
+		{
+			let sender = tx.sender().unwrap();
+			if let Some(_) = self.future.remove_by_address(&sender, &tx.nonce) {
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: `if self.future.remove_by_address(&sender, &tx.nonce).is_some()` might be slightly lighter on syntax?

File: sync/src/transaction_queue.rs [hunk 11]
```diff
@@ -0,0 +1,616 @@
...
+					self.future.insert(sender.clone(), v.tx.nonce, v);
+				}
+			}
+			self.future.enforce_limit();
+			return;
+		}
+
+		// Remove from future
+		{
+			let sender = tx.sender().unwrap();
```
Comment (gavofyork) [inline sync/src/transaction_queue.rs]: is this guaranteed not to panic?

File: util/src/table.rs [hunk 12]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
```
Comment (gavofyork) [inline util/src/table.rs]: this should be split out into a separate PR and submitted separately. should also include testing.

**Comment by gavofyork** [comment]
generally looks good, but it's a bit big. i think a fair amount of the changes (particularly `last_hashes` refactoring) have already been introduced in a separate PR. also the `Table` struct can be split out into a separate PR (along with tests).

what's left is better, but still probably > 500 line limit. ideally `TransactionQueue` would be introduced first along with its (fairly exhaustive) unit tests, and the integration would follow in a separate PR.

**Comment by NikVolf** [comment]
closing reopening this pr will help to reduce it, since some commits are already merged in other prs
